api/python/t4/data_transfer.py

```diff
@@ -607,3 +610,120 @@ def _process_url(src):
+
+    # use remaining file extensions to get format info, if none is present
+    if exts and not format:
+        ext = exts[-1].lower()
+        if ext in ext_formats:
+            if ext == '.tsv':
+                input_serialization = kwargs.setdefault('InputSerialization', {})
+                csv = input_serialization.setdefault('CSV', {})
+                if 'FieldDelimiter' not in csv:
+                    csv['FieldDelimiter'] = '\t'
+            format = ext_formats[ext]
+            s3_format = valid_s3_select_formats[format]
+            ok_compression = format_compression[s3_format]
+            if compression not in format_compression[s3_format]:
+                raise QuiltException("Compression {!r} not valid for select on format {!r}: "
+                                     "Expected {!r}".format(compression, s3_format, ok_compression))
+    if not format:
+        raise QuiltException("Unable to discover format for select on {!r}".format(url))
+
+    s3_format = valid_s3_select_formats[format]
+    input_serialization = kwargs.setdefault('InputSerialization', {})
+
+    if s3_format == 'JSON':
+        jsn = input_serialization.setdefault('JSON', {})
+        jsn['TYPE'] = "LINES" if format == 'jsonl' else "DOCUMENT"
+    elif s3_format == 'CSV':
+        csv = input_serialization.setdefault('CSV', {})
+        if format == 'tsv':
+            csv['FieldDelimiter'] = '\t'
+    elif s3_format == "Parquet":
+        input_serialization.setdefault('Parquet', {})
+
+    s3 = s3client if s3client else s3_client  # use passed in client, otherwise module-level client
+    if 'OutputSerialization' not in kwargs:
+        kwargs['OutputSerialization'] = {'JSON': {}}
+
+    response = s3.select_object_content(Bucket=bucket, Key=path, Expression=query, ExpressionType=query_type,
+                                        **kwargs)
+
+    # we don't want multiple copies of large chunks of data hanging around.
+    # ..iteration ftw.  It's what we get from amazon, anyways..
+    def iter_chunks(resp):
+        for item in resp['Payload']:
+            chunk = item.get('Records', {}).get('Payload')
+            if chunk is None:
+                continue
+            yield chunk
+
+    def iter_lines(resp, delimiter):
+        lastline = ''
+        for chunk in iterdecode(iter_chunks(resp), 'utf-8'):
+            lines = chunk.split(delimiter)
+            lines[0] = lastline + lines[0]
+            lastline = lines.pop(-1)
+            for line in lines:
+                yield line + delimiter
+        yield lastline
+
+    if not raw:
+        if 'JSON' in kwargs["OutputSerialization"]:
+            delimiter = kwargs['OutputSerialization']['JSON'].get('RecordDelimiter', '\n')
+            reader = jsonlines.Reader(line.strip() for line in iter_lines(response, delimiter)
+                                      if line.strip())
+            # noinspection PyPackageRequirements
+            from pandas import DataFrame
```

Comment:
Can you add a comment about why you're importing this here?

Reply:
Done.